src/execution_plan/optimizations/traverse_order.c

```diff
@@ -41,149 +26,68 @@ static inline void _Arrangement_Print(Arrangement arrangement, uint size) {
-	if((src->label || dest->label) &&
-	   AlgebraicExpression_Edge(exp) &&
-	   AlgebraicExpression_OperandCount(exp) == 1) return false;
-
-	for(int i = 1; i < exps_count; i++) {
-		exp = arrangement[i];
-		int j = i - 1;
-
-		// Scan previous expressions.
-		for(; j >= 0; j--) {
-			AlgebraicExpression *prev_exp = arrangement[j];
-			const char *exp_src = AlgebraicExpression_Source(exp);
-			const char *exp_dest = AlgebraicExpression_Destination(exp);
-			const char *prev_exp_src = AlgebraicExpression_Source(prev_exp);
-			const char *prev_exp_dest = AlgebraicExpression_Destination(prev_exp);
-
-			if(!RG_STRCMP(prev_exp_src, exp_src)     ||
-			   !RG_STRCMP(prev_exp_dest, exp_src)    ||
-			   !RG_STRCMP(prev_exp_src, exp_dest)    ||
-			   !RG_STRCMP(prev_exp_dest, exp_dest)) break;
-		}
-		/* Nither src or dest nodes are mentioned in previous expressions
-		 * as such the arrangement is invalid. */
-		if(j < 0) return false;
+	const char *src = AlgebraicExpression_Source(exp);
+	const char *dest = AlgebraicExpression_Destination(exp);
+	// Iterate over all previous expressions to make sure at least one of this expression's endpoints are bound.
+	for(int i = pos - 1; i >= 0; i --) {
+		AlgebraicExpression *prev = exps[i];
+		const char *prev_src = AlgebraicExpression_Source(prev);
+		const char *prev_dest = AlgebraicExpression_Destination(prev);
+		if(!RG_STRCMP(prev_src, src)     ||
+		   !RG_STRCMP(prev_dest, src)    ||
+		   !RG_STRCMP(prev_src, dest)    ||
+		   !RG_STRCMP(prev_dest, dest)) return true;
 	}
-	return true;
+	return false;
 }
 
-static int _penalty_arrangement(Arrangement arrangement, uint exp_count) {
-	int penalty = 0;
-	AlgebraicExpression *exp;
-
-	// Account for first expression transposes.
-	exp = arrangement[0];
-	uint transpose_count = AlgebraicExpression_OperationCount(exp, AL_EXP_TRANSPOSE);
-	penalty += transpose_count * T;
-
-	for(uint i = 1; i < exp_count; i++) {
-		exp = arrangement[i];
-		bool src_resolved = false;
+// Promote a selected operand to the given position while otherwise maintaining the optimal order.
+static void _promote_next_operand(AlgebraicExpression **exps, uint exp_count, uint to, uint from) {
+	AlgebraicExpression *tmp = exps[to];
+	// Migrate 'from' to 'to'
+	exps[to] = exps[from];
+	// Move all expressions between 'to' and 'from' 1 to the right.
+	for(uint i = to + 1; i < from; i ++) exps[i + 1] = exps[i];
+	// Set the replaced expression directly after its original position.
+	exps[to + 1] = tmp;
+}
 
-		// See if source is already resolved.
-		for(int j = i - 1; j >= 0; j--) {
-			AlgebraicExpression *prev_exp = arrangement[j];
-			if(!RG_STRCMP(AlgebraicExpression_Source(prev_exp), AlgebraicExpression_Source(exp)) ||
-			   !RG_STRCMP(AlgebraicExpression_Destination(prev_exp), AlgebraicExpression_Source(exp))) {
-				src_resolved = true;
+static void _order_expressions(AlgebraicExpression **exps, uint exp_count, QueryGraph *qg) {
```

Comment:
Once `_promote_next_operand` is called the order (top score at index 0, min score at the very end) of the array is tampered, as such we can no longer assume that when trying to positioning an item to position `i` that the first item we encounter (the item currently at position `i`) is the one with the highest score possible.

Consider the following score, element correspondence arrays
```
[80, 80, 40, 40, 20]
[F, F, L, L, E]
```
When trying to position the second item, we might end up with:
```
[80, 20, 40, 40, 80]
[F, E, L, L, F]
```

Assuming this sequence is valid, it will be accepted, but there might exists a better sequences, e.g.
```
[80, 20, 80, 40, 40]
[F, E, F, L, L]
```

I think it will be best if we take this one online.